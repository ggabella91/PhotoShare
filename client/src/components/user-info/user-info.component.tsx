import React, { useState } from 'react';
import { useHistory, NavLink } from 'react-router-dom';
import { connect } from 'react-redux';
import { Dispatch } from 'redux';
import { createStructuredSelector } from 'reselect';

import { AppState } from '../../redux/root-reducer';

import { DeleteReactionReq } from '../../redux/post/post.types';

import {
  setCommentToDelete,
  setShowCommentOptionsModal,
  setShowPostEditForm,
} from '../../redux/post/post.actions';

import './user-info.styles.scss';

export enum StyleType {
  suggestion = 'suggestion',
  modal = 'modal',
  feed = 'feed',
  comment = 'comment',
}

export interface UserInfoAndOtherData {
  profilePhotoFileString: string;
  username: string;
  name: string;
  location: string;
  comment: string;
  commentDate?: Date | string;
  reactionId?: string;
  reactingUserId?: string;
}

interface UserInfoProps {
  styleType: StyleType;
  userInfoArray: UserInfoAndOtherData[];
  isCaption?: boolean;
  isCaptionOwner?: boolean;
  setCommentToDelete: typeof setCommentToDelete;
  setShowCommentOptionsModal: typeof setShowCommentOptionsModal;
  setShowPostEditForm: typeof setShowPostEditForm;
}

export const UserInfo: React.FC<UserInfoProps> = ({
  userInfoArray,
  styleType,
  isCaption,
  isCaptionOwner,
  setCommentToDelete,
  setShowCommentOptionsModal,
  setShowPostEditForm,
}) => {
  const [showCommentOptionsButtonForIdx, setShowCommentOptionsButtonForIdx] =
    useState({ show: false, idx: -1 });

  let history = useHistory();

  const handleSetCommentToDelete = (idx: number) => {
    const commentToDelete = userInfoArray[idx];
    if (commentToDelete.reactionId && commentToDelete.reactingUserId) {
      setCommentToDelete({
        reactionId: commentToDelete.reactionId,
        isLikeRemoval: false,
        reactingUserId: commentToDelete.reactingUserId,
      });

      setShowCommentOptionsModal(true);
    }
  };

  const handleClickCaptionOptions = () => {
    if (isCaptionOwner) {
      setShowPostEditForm(true);
    }
  };

  const userInfo = userInfoArray.map(
    (el: UserInfoAndOtherData, idx: number) => (
      <div
        className={`user-${styleType}-element`}
        key={idx}
        onClick={
          styleType === StyleType.suggestion
            ? () => {
                history.push(`/${el.username}`);
              }
            : () => {}
        }
        onMouseEnter={() =>
          setShowCommentOptionsButtonForIdx({ show: true, idx })
        }
        onMouseLeave={() =>
          setShowCommentOptionsButtonForIdx({ show: false, idx })
        }
      >
        <div className={`${styleType}-avatar`}>
          {el.profilePhotoFileString ? (
            <img
              className={`${styleType}-profile-photo`}
              src={`data:image/jpeg;base64,${el.profilePhotoFileString}`}
              alt='profile-pic'
            />
          ) : null}
          {!el.profilePhotoFileString ? (
            <div className={`${styleType}-photo-placeholder`}>
              <span className={`${styleType}-photo-placeholder-text`}>
                No photo
              </span>
            </div>
          ) : null}
        </div>
        <div className='user-data-and-date'>
          <div className={`${styleType}-username-and-other-data`}>
            {styleType !== StyleType.suggestion ? (
              <NavLink
                to={`/${el.username}`}
                className={`${styleType}-username`}
              >
                {el.username}
              </NavLink>
            ) : (
              <span className={`${styleType}-username`}>{el.username}</span>
            )}
            <span className={`${styleType}-name`}>{el.name}</span>
            <span className={`${styleType}-location`}>{el.location}</span>
            <span>{el.comment ? el.comment : null}</span>
          </div>
          {el.commentDate ? (
            <span className={`${styleType}-date`}>
              {new Date(el.commentDate).toDateString()}
            </span>
          ) : null}
        </div>
        <div
          className={`${
            showCommentOptionsButtonForIdx.show &&
            showCommentOptionsButtonForIdx.idx === idx
              ? ''
              : 'hide'
          } comment-options`}
        >
          {styleType === StyleType.comment && !isCaption ? (
            <span
              className='comment-ellipsis-button'
              onClick={() => handleSetCommentToDelete(idx)}
            >
              ...
            </span>
          ) : null}
          {styleType === StyleType.comment && isCaption ? (
            <span
              className='comment-ellipsis-button'
              onClick={handleClickCaptionOptions}
            >
              ...
            </span>
          ) : null}
        </div>
      </div>
    )
  );

  return <div className={`user-${styleType}-container`}>{userInfo}</div>;
};

interface LinkStateProps {}

const mapStateToProps = createStructuredSelector<AppState, LinkStateProps>({});

const mapDispatchToProps = (dispatch: Dispatch) => ({
  setCommentToDelete: (deleteReactionReq: DeleteReactionReq) =>
    dispatch(setCommentToDelete(deleteReactionReq)),
  setShowCommentOptionsModal: (showCommentOptionsModal: boolean) =>
    dispatch(setShowCommentOptionsModal(showCommentOptionsModal)),
  setShowPostEditForm: (showPostEditForm: boolean) =>
    dispatch(setShowPostEditForm(showPostEditForm)),
});

export default connect(mapStateToProps, mapDispatchToProps)(UserInfo);
