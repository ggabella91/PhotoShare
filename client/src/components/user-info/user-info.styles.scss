.user-suggestion-container {
  display: flex;
  flex-direction: column;
  position: absolute;
  width: 220px;
  max-height: 100px;
  background-color: white;
  overflow-y: scroll;
  overflow-x: hidden;
  border-radius: 4px;
  transform: translate(-10px);
  z-index: 20;

  .user-suggestion-element {
    display: flex;
    position: relative;
    justify-content: start;
    width: 220px;
    height: 50px;
    background-color: white;
    border-bottom: solid 1px rgb(180, 180, 180);

    &:hover {
      cursor: pointer;
    }

    .suggestion-avatar {
      position: relative;
      width: 50px;
      height: 50px;
      border-radius: 50%;

      .suggestion-profile-photo {
        position: relative;
        margin-left: 50%;
        margin-top: 50%;
        transform: translate(-50%, -50%);
        width: 32px;
        height: 32px;
        border-radius: 50%;
      }

      .suggestion-photo-placeholder {
        position: relative;
        margin-left: 50%;
        margin-top: 50%;
        transform: translate(-50%, -50%);
        width: 32px;
        height: 32px;
        border-radius: 50%;
        border: solid 1px;
        border-color: black;
        display: flex;
        flex-direction: column;
        align-items: center;

        .suggestion-photo-placeholder-text {
          position: relative;
          font-size: 8px;
          color: black;
          margin-top: 50%;
          transform: translateY(-55%);
        }
      }
    }

    .suggestion-username-and-other-data {
      display: flex;
      position: relative;
      flex-direction: column;
      padding-top: 4px;
      width: 140px;
      height: 50px;
      justify-self: flex-end;
      text-align: left;

      .suggestion-username {
        display: flex;
        position: relative;
        font-size: 14px;
        font-weight: 420;
        color: black;
      }

      .suggestion-name {
        display: flex;
        position: relative;
        font-size: 12px;
        font-weight: 380;
        color: black;
      }

      .suggestion-location {
        display: none;
      }
    }
  }
}

.user-modal-container {
  display: flex;
  flex-direction: column;
  position: relative;
  width: inherit;
  max-height: 338px;
  background-color: white;
  border-radius: 4px;

  .user-modal-element {
    display: flex;
    position: relative;
    justify-content: start;
    width: 100%;
    height: 70px;
    background-color: white;
    border-bottom: solid 1px rgb(180, 180, 180);
    padding-top: 5px;
    padding-bottom: 5px;

    &:hover {
      cursor: pointer;
    }

    .modal-avatar {
      position: relative;
      width: 60px;
      height: 60px;
      border-radius: 50%;
      margin-left: 15px;
      margin-right: 2vw;

      .modal-profile-photo {
        position: relative;
        margin-left: 50%;
        margin-top: 50%;
        transform: translate(-50%, -50%);
        width: 45px;
        height: 45px;
        border-radius: 50%;
      }

      .modal-photo-placeholder {
        position: relative;
        margin-left: 50%;
        margin-top: 50%;
        transform: translate(-50%, -50%);
        width: 45px;
        height: 45px;
        border-radius: 50%;
        border: solid 1px;
        border-color: black;
        display: flex;
        flex-direction: column;
        align-items: center;
        justify-content: center;

        .modal-photo-placeholder-text {
          position: relative;
          font-size: 10px;
          color: black;
          margin-top: 50%;
          margin-left: 50%;
          transform: translate(-50%, -50%);
        }
      }
    }

    .modal-username-and-other-data {
      display: flex;
      position: relative;
      flex-direction: column;
      padding-top: 6px;
      width: 75%;
      height: 60px;
      justify-self: flex-end;
      text-align: left;

      .modal-username {
        display: flex;
        position: relative;
        font-size: 16px;
        font-weight: 420;
        color: black;
      }

      .modal-name {
        display: flex;
        position: relative;
        font-size: 14px;
        font-weight: 380;
        color: black;
      }

      .modal-location {
        display: none;
      }
    }
  }
}

.user-feed-container {
  display: flex;
  flex-direction: column;
  position: relative;
  width: inherit;
  max-height: 338px;
  background-color: white;
  border-radius: 4px;

  .user-feed-element {
    display: flex;
    position: relative;
    justify-content: start;
    width: 100%;
    height: 60px;
    background-color: white;
    border-bottom: solid 1px rgb(180, 180, 180);
    padding-top: 5px;
    padding-bottom: 5px;

    &:hover {
      cursor: pointer;
    }

    .feed-avatar {
      position: relative;
      width: 50px;
      height: 50px;
      border-radius: 50%;
      margin-left: 15px;
      margin-right: 15px;
      padding-top: 5px;

      .feed-profile-photo {
        position: relative;
        margin-left: 50%;
        margin-top: 50%;
        transform: translate(-50%, -50%);
        width: 40px;
        height: 40px;
        border-radius: 50%;
      }

      .feed-photo-placeholder {
        position: relative;
        margin-left: 50%;
        margin-top: 50%;
        transform: translate(-50%, -50%);
        width: 40px;
        height: 40px;
        border-radius: 50%;
        border: solid 1px;
        border-color: black;
        display: flex;
        flex-direction: column;
        align-items: center;
        justify-content: center;

        .feed-photo-placeholder-text {
          position: relative;
          font-size: 10px;
          color: black;
          margin-top: 50%;
          margin-left: 50%;
          transform: translate(-50%, -50%);
        }
      }
    }

    .feed-username-and-other-data {
      display: flex;
      position: relative;
      flex-direction: column;
      padding-top: 6px;
      width: 75%;
      height: 60px;
      justify-self: flex-end;
      text-align: left;

      .feed-username {
        display: flex;
        position: relative;
        font-size: 14px;
        font-weight: 540;
        color: black;
      }

      .feed-name {
        display: none;
      }

      .feed-location {
        display: flex;
        position: relative;
        font-size: 12px;
        font-weight: 380;
        color: black;
      }
    }
  }
}

.user-comment-container {
  display: flex;
  flex-direction: column;
  position: relative;
  width: inherit;
  max-height: 338px;
  background-color: white;
  border-radius: 4px;

  .user-comment-element {
    display: flex;
    position: relative;
    justify-content: start;
    width: 100%;
    height: 60px;
    background-color: white;
    border-bottom: solid 1px rgb(180, 180, 180);
    padding-top: 5px;
    padding-bottom: 5px;

    &:hover {
      cursor: pointer;
    }

    .comment-avatar {
      position: relative;
      width: 50px;
      height: 50px;
      border-radius: 50%;
      margin-left: 15px;
      margin-right: 15px;
      padding-top: 5px;

      .comment-profile-photo {
        position: relative;
        margin-left: 50%;
        margin-top: 50%;
        transform: translate(-50%, -50%);
        width: 40px;
        height: 40px;
        border-radius: 50%;
      }

      .comment-photo-placeholder {
        position: relative;
        margin-left: 50%;
        margin-top: 50%;
        transform: translate(-50%, -50%);
        width: 40px;
        height: 40px;
        border-radius: 50%;
        border: solid 1px;
        border-color: black;
        display: flex;
        flex-direction: column;
        align-items: center;
        justify-content: center;

        .comment-photo-placeholder-text {
          position: relative;
          font-size: 10px;
          color: black;
          margin-top: 50%;
          margin-left: 50%;
          transform: translate(-50%, -50%);
        }
      }
    }

    .comment-username-and-other-data {
      display: flex;
      position: relative;
      flex-direction: column;
      padding-top: 6px;
      width: 75%;
      height: 60px;
      justify-self: flex-end;
      text-align: left;

      .comment-username {
        display: flex;
        position: relative;
        font-size: 14px;
        font-weight: 540;
        color: black;
      }

      .comment-name {
        display: none;
      }

      .comment-location {
        display: none;
      }

      .comment-text {
        display: flex;
        position: relative;
        font-size: 12;
        font-weight: 400;
      }
    }

    .comment-date {
      display: flex;
      position: relative;
      font-size: 10;
      font-weight: 380;
      margin-top: 5px;
    }
  }
}
