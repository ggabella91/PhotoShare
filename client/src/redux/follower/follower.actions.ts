import {
  Follower,
  FollowError,
  UsersFollowingRequest,
  FollowerActions,
  FollowerActionTypes,
  NotificationActions,
  NotificationActionTypes,
  PostNotificationReq,
  Notification,
  NotificationError,
  GetNotificationsReq,
} from './follower.types';

export const followNewUserStart = (
  userToFollowId: string
): FollowerActionTypes => ({
  type: FollowerActions.FOLLOW_NEW_USER_START,
  payload: userToFollowId,
});

export const followNewUserSuccess = (message: string): FollowerActionTypes => ({
  type: FollowerActions.FOLLOW_NEW_USER_SUCCESS,
  payload: message,
});

export const followNewUserFailure = (
  error: FollowError
): FollowerActionTypes => ({
  type: FollowerActions.FOLLOW_NEW_USER_FAILURE,
  payload: error,
});

export const getFollowersStart = (userId: string): FollowerActionTypes => ({
  type: FollowerActions.GET_FOLLOWERS_START,
  payload: userId,
});

export const getFollowersSuccess = (
  followers: Follower[]
): FollowerActionTypes => ({
  type: FollowerActions.GET_FOLLOWERS_SUCCESS,
  payload: followers,
});

export const getFollowersFailure = (
  error: FollowError
): FollowerActionTypes => ({
  type: FollowerActions.GET_FOLLOWERS_FAILURE,
  payload: error,
});

export const getUsersFollowingStart = ({
  userId,
  whoseUsersFollowing,
}: UsersFollowingRequest): FollowerActionTypes => ({
  type: FollowerActions.GET_USERS_FOLLOWING_START,
  payload: {
    userId,
    whoseUsersFollowing,
  },
});

export const getCurrentUserUsersFollowingSuccess = (
  usersFollowing: Follower[]
): FollowerActionTypes => ({
  type: FollowerActions.GET_CURRENT_USER_USERS_FOLLOWING_SUCCESS,
  payload: usersFollowing,
});

export const getOtherUserUsersFollowingSuccess = (
  usersFollowing: Follower[]
): FollowerActionTypes => ({
  type: FollowerActions.GET_OTHER_USER_USERS_FOLLOWING_SUCCESS,
  payload: usersFollowing,
});

export const getUsersFollowingFailure = (
  error: FollowError
): FollowerActionTypes => ({
  type: FollowerActions.GET_USERS_FOLLOWING_FAILURE,
  payload: error,
});

export const unfollowUserStart = (userId: string): FollowerActionTypes => ({
  type: FollowerActions.UNFOLLOW_USER_START,
  payload: userId,
});

export const unfollowUserSuccess = (message: string): FollowerActionTypes => ({
  type: FollowerActions.UNFOLLOW_USER_SUCCESS,
  payload: message,
});

export const unfollowUserFailure = (
  error: FollowError
): FollowerActionTypes => ({
  type: FollowerActions.UNFOLLOW_USER_FAILURE,
  payload: error,
});

export const clearFollowState = (): FollowerActionTypes => ({
  type: FollowerActions.CLEAR_FOLLOW_STATE,
});

export const postNotificationStart = (
  postNotificationReq: PostNotificationReq
): NotificationActionTypes => ({
  type: NotificationActions.POST_NOTIFICATION_START,
  payload: postNotificationReq,
});

export const postNotificationSuccess = (
  notification: Notification
): NotificationActionTypes => ({
  type: NotificationActions.POST_NOTIFICATION_SUCCESS,
  payload: notification,
});

export const postNotificationFailure = (
  error: NotificationError
): NotificationActionTypes => ({
  type: NotificationActions.POST_NOTIFICATION_FAILURE,
  payload: error,
});

export const getNotificationsStart = (
  getNotificationsReq: GetNotificationsReq
): NotificationActionTypes => ({
  type: NotificationActions.GET_NOTIFICATIONS_START,
  payload: getNotificationsReq,
});

export const setNotificationsQueryLength = (
  queryLength: number
): NotificationActionTypes => ({
  type: NotificationActions.SET_NOTIFICATIONS_QUERY_LENGTH,
  payload: queryLength,
});

export const getNotificationsSuccess = (
  notifications: Notification[]
): NotificationActionTypes => ({
  type: NotificationActions.GET_NOTIFICATIONS_SUCCESS,
  payload: notifications,
});

export const getNotificationsFailure = (
  error: NotificationError
): NotificationActionTypes => ({
  type: NotificationActions.GET_NOTIFICATIONS_FAILURE,
  payload: error,
});
