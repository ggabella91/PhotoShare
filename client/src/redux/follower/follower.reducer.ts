import {
  FollowerActions,
  FollowerActionTypes,
  FollowerState,
} from './follower.types';

const INITIAL_STATE: FollowerState = {
  followers: null,
  currentUserUsersFollowing: null,
  otherUserUsersFollowing: null,
  followConfirm: null,
  followError: null,
  getFollowersConfirm: null,
  getFollowersError: null,
  getUsersFollowingConfirm: null,
  getUsersFollowingError: null,
  unfollowConfirm: null,
  unfollowError: null,
};

const followerReducer = (
  state = INITIAL_STATE,
  action: FollowerActionTypes
) => {
  switch (action.type) {
    case FollowerActions.FOLLOW_NEW_USER_SUCCESS:
      return {
        ...state,
        followConfirm: 'New user followed!',
        followError: null,
      };
    case FollowerActions.GET_FOLLOWERS_SUCCESS:
      return {
        ...state,
        followers: action.payload,
        getFollowersConfirm: 'Followers fetched successfully!',
        getFollowersError: null,
      };
    case FollowerActions.GET_CURRENT_USER_USERS_FOLLOWING_SUCCESS:
      return {
        ...state,
        currentUserUsersFollowing: action.payload,
        getUsersFollowingConfirm:
          "Current user's users following fetched successfully!",
        getUsersFollowingError: null,
      };
    case FollowerActions.GET_OTHER_USER_USERS_FOLLOWING_SUCCESS:
      return {
        ...state,
        otherUserUsersFollowing: action.payload,
        getUsersFollowingConfirm:
          "Other user's users following fetched successfully!",
        getUsersFollowingError: null,
      };
    case FollowerActions.UNFOLLOW_USER_SUCCESS:
      return {
        ...state,
        unfollowConfirm: action.payload,
        unfollowError: null,
      };
    case FollowerActions.FOLLOW_NEW_USER_FAILURE:
      return {
        ...state,
        followConfirm: null,
        followError: action.payload,
      };
    case FollowerActions.GET_FOLLOWERS_FAILURE:
      return {
        ...state,
        getFollowersConfirm: null,
        getFollowersError: action.payload,
      };
    case FollowerActions.GET_USERS_FOLLOWING_FAILURE:
      return {
        ...state,
        getUsersFollowingConfirm: null,
        getUsersFollowingError: action.payload,
      };
    case FollowerActions.UNFOLLOW_USER_FAILURE:
      return {
        ...state,
        unfollowConfirm: null,
        unfollowError: action.payload,
      };
    case FollowerActions.CLEAR_UNFOLLOW_PROPERTIES:
      return {
        ...state,
        unfollowConfirm: null,
        unfollowError: null,
      };
    default:
      return state;
  }
};

export default followerReducer;
