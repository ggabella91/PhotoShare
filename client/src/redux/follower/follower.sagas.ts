import { takeLatest, put, all, call } from 'redux-saga/effects';
import { SagaIterator } from '@redux-saga/core';
import { ActionPattern, Saga } from '@redux-saga/types';

import {
  Follower,
  FollowError,
  WhoseUsersFollowing,
  UsersFollowingRequest,
  FollowerActions,
  FollowerActionTypes,
} from './follower.types';

import {
  followNewUserSuccess,
  followNewUserFailure,
  getFollowersSuccess,
  getFollowersFailure,
  getUsersFollowingStart,
  getCurrentUserUsersFollowingSuccess,
  getOtherUserUsersFollowingSuccess,
  getUsersFollowingFailure,
} from './follower.actions';

import axios from 'axios';

export function* followNewUser({
  payload: userToFollowId,
}: {
  payload: string;
}): any {
  try {
    const { data } = yield axios.post(
      `/api/followers/follow-new-user/${userToFollowId}`
    );

    yield put(followNewUserSuccess(data));
  } catch (err) {
    yield put(followNewUserFailure(err));
  }
}

export function* getFollowers({
  payload: userToFollowId,
}: {
  payload: string;
}): any {
  try {
    const { data } = yield axios.get(
      `/api/followers/get-followers/${userToFollowId}`
    );

    yield put(getFollowersSuccess(data));
  } catch (err) {
    yield put(getFollowersFailure(err));
  }
}

export function* getUsersFollowing({
  payload: { userId, whoseUsersFollowing },
}: {
  payload: UsersFollowingRequest;
}): any {
  try {
    const { data } = yield axios.get(
      `/api/followers/get-users-following/${userId}`
    );

    if (whoseUsersFollowing === WhoseUsersFollowing.CURRENT_USER) {
      yield put(getCurrentUserUsersFollowingSuccess(data));
    } else if (whoseUsersFollowing === WhoseUsersFollowing.OTHER_USER) {
      yield put(getOtherUserUsersFollowingSuccess(data));
    }
  } catch (err) {
    yield put(getUsersFollowingFailure(err));
  }
}

export function* onFollowNewUserStart(): SagaIterator {
  yield takeLatest<ActionPattern, Saga>(
    FollowerActions.FOLLOW_NEW_USER_START,
    followNewUser
  );
}

export function* onGetFollowersStart(): SagaIterator {
  yield takeLatest<ActionPattern, Saga>(
    FollowerActions.GET_FOLLOWERS_START,
    getFollowers
  );
}

export function* onGetUsersFollowingStart(): SagaIterator {
  yield takeLatest<ActionPattern, Saga>(
    FollowerActions.GET_USERS_FOLLOWING_START,
    getUsersFollowing
  );
}

export function* followerSagas(): SagaIterator {
  yield all([
    call(onFollowNewUserStart),
    call(onGetFollowersStart),
    call(onGetUsersFollowingStart),
  ]);
}
