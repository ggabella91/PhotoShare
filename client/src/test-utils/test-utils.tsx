import { FC, ReactElement } from 'react';
import { BrowserRouter, Routes, Route } from 'react-router-dom';
import { Provider } from 'react-redux';
import { render, RenderOptions } from '@testing-library/react';
import { store } from '../redux/store';

interface WrapperProps {
  wrapperProps?: { route: string; location: string };
}

type CustomRenderOptions = RenderOptions & WrapperProps;

const WithReduxProvider: FC<WrapperProps> = ({ children, wrapperProps }) => {
  return (
    <Provider store={store}>
      <BrowserRouter>
        <Routes>
          <Route path={wrapperProps?.route || '/'} element={children} />
        </Routes>
      </BrowserRouter>
    </Provider>
  );
};

const renderWithReduxAndRouter = (
  ui: ReactElement,
  options?: Omit<CustomRenderOptions, 'wrapper'>
) => {
  window.history.pushState({}, '', options?.wrapperProps?.location || '/');
  const customContainer = document.createElement('div');
  customContainer.id = 'root';

  return render(ui, {
    wrapper: (props) => (
      <WithReduxProvider {...props} wrapperProps={options?.wrapperProps} />
    ),
    container: document.body.appendChild(customContainer),
    ...options,
  });
};

export * from '@testing-library/react';
export { default as userEvent } from '@testing-library/user-event';
export { renderWithReduxAndRouter as render };
