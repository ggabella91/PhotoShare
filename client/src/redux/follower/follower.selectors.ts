import { createSelector } from 'reselect';

import { AppState } from '../root-reducer';
import { FollowerState } from './follower.types';

const selectFollowerState = (state: AppState): FollowerState => state.follower;

export const selectFollowers = createSelector(
  [selectFollowerState],
  (followerState: FollowerState) => followerState.followers
);

export const selectCurrentUserUsersFollowing = createSelector(
  [selectFollowerState],
  (followerState: FollowerState) => followerState.currentUserUsersFollowing
);

export const selectOtherUserUsersFollowing = createSelector(
  [selectFollowerState],
  (followerState: FollowerState) => followerState.otherUserUsersFollowing
);

export const selectFollowConfirm = createSelector(
  [selectFollowerState],
  (followerState: FollowerState) => followerState.followConfirm
);

export const selectFollowError = createSelector(
  [selectFollowerState],
  (followerState: FollowerState) => followerState.followError
);

export const selectGetFollowersConfirm = createSelector(
  [selectFollowerState],
  (followerState: FollowerState) => followerState.getFollowersConfirm
);

export const selectGetFollowersError = createSelector(
  [selectFollowerState],
  (followerState: FollowerState) => followerState.getFollowersError
);

export const selectGetUsersFollowingConfirm = createSelector(
  [selectFollowerState],
  (followerState: FollowerState) => followerState.getUsersFollowingConfirm
);

export const selectGetUsersFollowingError = createSelector(
  [selectFollowerState],
  (followerState: FollowerState) => followerState.getUsersFollowingError
);

export const selectUnfollowConfirm = createSelector(
  [selectFollowerState],
  (followerState: FollowerState) => followerState.unfollowConfirm
);

export const selectUnfollowError = createSelector(
  [selectFollowerState],
  (followerState: FollowerState) => followerState.unfollowError
);

export const selectPostNotificationConfirm = createSelector(
  [selectFollowerState],
  (followerState: FollowerState) => followerState.postNotificationConfirm
);

export const selectPostNotificationError = createSelector(
  [selectFollowerState],
  (followerState: FollowerState) => followerState.postNotificationError
);

export const selectNotifications = createSelector(
  [selectFollowerState],
  (followerState: FollowerState) => followerState.notifications
);

export const selectGetNotificationsError = createSelector(
  [selectFollowerState],
  (followerState: FollowerState) => followerState.getNotificationsError
);
