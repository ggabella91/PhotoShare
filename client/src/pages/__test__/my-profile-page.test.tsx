import { render, screen } from '../../test-utils/test-utils';
import { MyProfilePage } from '../my-profile/my-profile-page.component';

import { Post, PostFile, DeleteReactionReq } from '../../redux/post/post.types';

import { User } from '../../redux/user/user.types';

import { Follower } from '../../redux/follower/follower.types';

import { UserInfoAndOtherData } from '../../components/user-info/user-info.component';

describe('my-profile-page component tests', () => {
  const setup = () => {
    const currentUser = {} as User;
    const followers = [{}, {}] as Follower[];
    const postData = [{}, {}, {}, {}] as Post[];
    const postFiles = [{}, {}, {}, {}] as PostFile[];
    const currentUserUsersFollowing = [{}, {}] as Follower[];
    const commentToDelete = {} as DeleteReactionReq;
    const postLikingUsersArray = [{}, {}] as UserInfoAndOtherData[];

    const getPostDataStart = jest.fn();
    const getPostFileStart = jest.fn();
    const archivePostStart = jest.fn();
    const clearArchivePostStatuses = jest.fn();
    const clearPostState = jest.fn();
    const getFollowersStart = jest.fn();
    const clearFollowersAndFollowing = jest.fn();
    const clearFollowPhotoFileArray = jest.fn();
    const getUsersFollowingStart = jest.fn();
    const clearFollowState = jest.fn();
    const setIsCurrentUserProfilePage = jest.fn();
    const setShowCommentOptionsModal = jest.fn();
    const deleteReactionStart = jest.fn();
    const setShowPostEditForm = jest.fn();

    render(
      <MyProfilePage
        currentUser={currentUser}
        profilePhotoKey='test-photo-filestring'
        profilePhotoFile={null}
        postFiles={postFiles}
        postError={null}
        getPostDataStart={getPostDataStart}
        getPostDataConfirm='confirm'
        getPostDataError={null}
        getPostFileStart={getPostFileStart}
        getPostFileConfirm='confirm'
        getPostFileError={null}
        archivePostStart={archivePostStart}
        archivePostConfirm={null}
        archivePostError={null}
        followers={followers}
        currentUserUsersFollowing={currentUserUsersFollowing}
        commentToDelete={commentToDelete}
        showCommentOptionsModal={false}
        postLikingUsersArray={postLikingUsersArray}
        getSinglePostDataConfirm={null}
        clearArchivePostStatuses={clearArchivePostStatuses}
        clearPostState={clearPostState}
        clearFollowPhotoFileArray={clearFollowPhotoFileArray}
        getFollowersStart={getFollowersStart}
        getUsersFollowingStart={getUsersFollowingStart}
        clearFollowersAndFollowing={clearFollowersAndFollowing}
        clearFollowState={clearFollowState}
        setIsCurrentUserProfilePage={setIsCurrentUserProfilePage}
        setShowCommentOptionsModal={setShowCommentOptionsModal}
        deleteReactionStart={deleteReactionStart}
        setShowPostEditForm={setShowPostEditForm}
      />
    );

    return {
      getPostDataStart,
      getPostFileStart,
      archivePostStart,
      clearArchivePostStatuses,
      clearPostState,
      getFollowersStart,
      clearFollowersAndFollowing,
      clearFollowPhotoFileArray,
      getUsersFollowingStart,
      clearFollowState,
      setIsCurrentUserProfilePage,
      setShowCommentOptionsModal,
      deleteReactionStart,
      setShowPostEditForm,
    };
  };

  it('renders a my-profile-page component', () => {
    const {
      getPostDataStart,
      getPostFileStart,
      getFollowersStart,
      getUsersFollowingStart,
    } = setup();

    const myProfilePage = screen.getByTestId('my-profile-page');

    expect(myProfilePage).toBeInTheDocument();
    expect(getPostDataStart).toBeCalled();
    expect(getPostFileStart).toBeCalled();
    expect(getFollowersStart).toBeCalled();
    expect(getUsersFollowingStart).toBeCalled();
  });
});
