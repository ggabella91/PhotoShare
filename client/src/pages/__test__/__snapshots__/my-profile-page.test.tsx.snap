// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a my-profile-page component 1`] = `
<div
  className="profile-page"
>
  <div
    className="user-bio"
  >
    <div
      className="avatar-and-details"
    >
      <div
        className="avatar"
      >
        <div
          className="user-bio-photo-placeholder"
        >
          <span
            className="user-bio-photo-placeholder-text"
          >
            No photo
          </span>
        </div>
      </div>
      <div
        className="user-details"
      >
        <div
          className="username-and-edit"
        >
          <span
            className="user-username"
          />
          <NavLink
            className="edit-profile"
            to="/settings"
          >
            <span
              className="edit-text"
            >
              Edit profile
            </span>
          </NavLink>
        </div>
        <NavLink
          className="edit-profile-narrow-screen"
          to="/settings"
        >
          <span
            className="edit-narrow-text"
          >
            Edit profile
          </span>
        </NavLink>
        <div
          className="posts-followers-following-stats"
        >
          <span
            className="user-stat"
          >
            0
             Posts
          </span>
          <span
            className="user-stat"
            onClick={[Function]}
          >
            0
             Followers
          </span>
          <span
            className="user-stat"
            onClick={[Function]}
          >
            0
             Following
          </span>
        </div>
        <div
          className="name-and-bio"
        >
          <span
            className="user-name"
          />
          <span
            className="user-bio"
          />
        </div>
      </div>
    </div>
    <div
      className="name-and-bio-narrow-screen"
    >
      <span
        className="user-name-narrow"
      />
      <span
        className="user-bio-narrow"
      />
    </div>
    <div
      className="posts-followers-following-stats-narrow-screen"
    >
      <ul
        className="stats-list"
      >
        <li
          className="stats-item"
        >
          0
           Posts
        </li>
        <li
          className="stats-item"
          onClick={[Function]}
        >
          0
           Followers
        </li>
        <li
          className="stats-item"
          onClick={[Function]}
        >
          0
           Following
        </li>
      </ul>
    </div>
  </div>
  <div
    className="posts-grid"
  />
  <Connect(PostModal)
    caption=""
    createdAt={2021-01-09T22:39:39.945Z}
    fileString=""
    location=""
    onHide={[Function]}
    onOptionsClick={[Function]}
    postId=""
    show={false}
    userId=""
    userName=""
    userProfilePhotoFile=""
  />
  <PostOptionsModal
    archive={[Function]}
    isCurrentUserPost={true}
    onHide={[Function]}
    show={false}
  />
  <Connect(FollowersOrFollowingModal)
    isFollowersModal={true}
    onHide={[Function]}
    show={false}
    users={null}
  />
</div>
`;
